### ISSUE TITLE
Regression: no rich text links work inside Streamfields on service page steps or guide pages. 

### ISSUE THREAD

**Issue by easherma** [cityofaustin/techstack#3852]
Here's a nice regression that i discovered while working on #3682
https://github.com/cityofaustin/techstack/issues/3053
( https://github.com/cityofaustin/joplin/pull/420 )
Did its thing, QA'ed, merged
https://github.com/cityofaustin/techstack/issues/3504
(https://github.com/cityofaustin/joplin/pull/499)
Did its thing, code change merged in actually undoes the fix of the previous commit

So you might notice that internal links inside rich text fields in streamfields don't work anymore, but I don't think we noticed at the time that the new code merged in was undoing a previous fix and the new code wasn't incorporating the old code :tada:

Hoping to sort this out as part of the changes for #3682, but we should probably also have a regression test to cover cases like this (in progress here: https://github.com/cityofaustin/joplin/pull/545)

Two main goals:
* fix the bug
* have a test plan for making sure the bug doesn't happen again in the future

**Comment by easherma** [comment]
Ways this could be tested: 
Unit tests for schema.py in joplin
integration (or maybe a unit test?) for the component in jans? Does every link navigate correctly? Does a link have an href?

**Comment by easherma** [comment]
https://github.com/syrusakbary/snapshottest/
https://docs.graphene-python.org/en/latest/testing/
https://docs.graphene-python.org/projects/django/en/latest/testing/

**Comment by desigonz** [comment]
@desigonz to write up issue for the regression + test case with acceptance criteria

**Comment by easherma** [comment]
Braindump for places to test:

Unit tests:
For model/page:

- Does the rich text widget have the expected list of features? ( not super worth testing)
- Do the links inside the author interface go to the right place on janis ( they do not, we could write a failing test, making this actually work would be its own work )

For the API (kinda systems test?):
- Test cases for our Streamfields interface ( would love input on fleshing this out in useful ways) 
- Test if an expected query matches output
    - For all instances that have links to expand, are they expanded? (can check the structure)
    - If expanded to a link, does the link work (this might be better off as a janis test)
    - Can we have a test that checks if we are checking in new code that changes the graphql output? (i.e. how could we have spotted the regression or at least noted that we were making a change)

**Comment by easherma** [comment]
changed the estimate as the original estimate was just to document the regression test, but we really need to be able to fix the problem to unblock #3682 and future features

**Comment by easherma** [comment]
Ideal test case:
Every rich text field has an internal or doc link
ServicePage with steps
steps include a steps_with_location block
There is also at least once component included from 'dynamicContent'
There are also several steps with options, which have nested options (which are rich text and should have a test internal link field or document link field)

**Comment by easherma** [comment]
https://github.com/cityofaustin/joplin/tree/3567-api-rep-error <--- strips out the custom service page stuff, fixes some errors, dosen't display locations stuff the way we want, but we could handle that use case

https://github.com/cityofaustin/joplin/pull/541 <--- leaves more code in place, sorta works, dosen't work for nested streamfields (such as options)

There might be a way for me to intersect these solutions, otherwise at some point it'll make sense to abandon one.

**Comment by briaguya** [comment]
left some notes on https://github.com/cityofaustin/joplin/pull/555, I think it's an ok route to go for now but I'd like to build out a pattern that allows for queryable streamfield blocks instead of just hardcoding our data model

I'm ok with this (1 - strips out the custom service page stuff, fixes some errors, has related location info in pretty much the same structure) as a temporary solution, if we also have a Janis PR out for the steps with location api changes, and make a new issue in the backlog for queryable streamfields

**Comment by easherma** [comment]
Ready-ish to test @:
https://janis-3567-richtext-links.netlify.com

https://joplin-pr-3567-api-rep-error.herokuapp.com/

**Comment by desigonz** [comment]
## Test cases

### Service page
- [ ] In Joplin, user adds internal link within **basic step** and publishes page; on Janis, the link opens to the right page
- [ ] In Joplin, user adds internal link within **step with options** and publishes page; on Janis, the link opens to the right page
- [ ] In Joplin, user adds internal link within **step with location** and publishes page; on Janis, the link opens to the right page
- [ ] In Joplin, user adds internal link within **additional content rich text block** and publishes page; on Janis, the link opens to the right page
### Information page
- [ ] In Joplin, user adds internal link within **rich text box** and publishes page; on Janis, the link opens to the right page
### Guide page
- [ ] In Joplin, user publishes a guide with the service page and information page tested above (with internal links); on Janis, the links open to the right page

**Comment by desigonz** [comment]
I added links to and published the follow pages on Joplin:
- [Service page](http://joplin-pr-3567-api-rep-error.herokuapp.com/admin/pages/363/edit/#tab-content)
- [Info page](http://joplin-pr-3567-api-rep-error.herokuapp.com/admin/pages/364/edit/)
- [Guide](http://joplin-pr-3567-api-rep-error.herokuapp.com/admin/pages/365/edit/)
But when I click "view live" in Joplin, the pages 404. I logged into Netlify and the last build was at 9:50 this morning:
![image](https://user-images.githubusercontent.com/5839508/74271066-e873cb00-4cd1-11ea-8f3b-6ba4bbc5a27b.png)

Can you look into how the Janis and Joplin branches are hooked up?

**Comment by desigonz** [comment]
I was able to trigger the builds via Netlify and test out the cases above. 

QA passed for the test cases! Looks and works great. We can spin out writing unit tests into a different issue, but use those test cases as the basis for those unit tests.

**Comment by desigonz** [comment]
@easherma Content has been sharing a lot of internal links that aren't working on published pages, such as [this one](https://alpha.austin.gov/en/permits-tickets/business-permits-and-licenses/food-business-permits/renew-your-mobile-food-vendor-permit/) on the mobile food vendor guide. Do you have an estimate on when this will be merged? Weighing whether to edit the pages with external links or waiting on this fix to be in production.
